files/scripts/supervisor-proc-exit-listener

```diff
@@ -70,9 +175,6 @@ def main(argv):
         headers = childutils.get_headers(line)
         payload = sys.stdin.read(int(headers['len']))
 
-        # Transition from READY to ACKNOWLEDGED
```

Comment:
use select on line 174.

https://repolinux.wordpress.com/2012/10/09/non-blocking-read-from-stdin-in-python/

Reply:
I tried the `select` method of non-blocking I/O. If we assigned the `0` to the fourth argument of `select` method (https://docs.python.org/3/library/select.html#select.select), it will continue do the polling the input and never block. That means the `while` loop from line 157 to 200 will running forever.

Reply:
I also tried another method. Actually if a critical process exited and event listener received this, it will determine whether the auto-restart feature was enabled or not. If auto-restart feature was enabled, then container will be restarted. Otherwise, we can fork a separate thread to run the function `monitoring_process_and_alerting(...)` to generate the alerting message and then event listener can go back to listen.

Reply:
Regarding `select`, if you set the timeout to zero, if will never block. You could set the value to zero or 1 second; either way you would need to check if anything was received on stdin. If so you would process the message. If not, you would check all previously exited processes to see if they have come back. If not, you would check if the alerting interval has elapsed and if so, log a new alert.

The threading alternative would also work, but would add complexity.

Reply:
I did some investigation about using the method `select(...)` last week. If we set the timeout to `0` and also there was no event sending to the listener, then the `while` loop between the line 170 ~ 200 will spin continuously and CPU usage of SwSS was more than 100%. This was not acceptable. 

If we set the timeout value to non-zero value such as 1 second or 5 seconds, according to your suggestion, we will not use the function `monitoring_process_and_alerting(...)` at line 118 to do monitoring and alerting. Then the alerting logic will be as following:

// Record when the event listener did the alerting of a process
under_monitoring = {}    
while True:                         
---- listener becomes ready.

@jleveque  Can you help me check whether my understanding was correct or not please?
---- stdin_descriptors = select.select([sys.stdin], [], [], **timeout**)[0]

---- for process in under_monitoring:
-------- if process becomes running:
------------ remove_from_alerting
-------- else if current_time - under_monitoring[process] == alerting_interval:
------------ alerting_and_update_timestamp

---- if stdin_descriptors:
-------- for descriptor in stdin_descriptors:
------------ line = descriptor.readline()
------------ parse_the_line
------------ if event_type == "EXITED":
---------------- if auto-restart was enabled:
-------------------- restart_container
---------------- else:
-------------------- alerting_and_update_timestamp

---- listener becomes ok.